spec/src/main/asciidoc/contextproviders.asciidoc

```diff
@@ -0,0 +1,129 @@
+
+With this model, the provider of thread context implements the org.eclipse.microprofile.concurrent.spi.ThreadContextProvider interface and packages it in a way that makes it available to the ServiceLoader. ThreadContextProvider identifies the thread context type and provides a way to capture snapshots of thread context as well as for applying empty/cleared context to threads.
+
+
+=== Example
+
+The following is a working example of a thread context provider and related interfaces.
+The example context type that it propagates is the priority of a thread. This is chosen, not because it is useful in any way, but because the concept of thread priority is simple, well understood, and already built into Java, allowing the reader to focus on the mechanisms of thread context capture/propagate/restore rather than the details of the context type itself.
+
+==== ThreadContextProvider
+
+The interface, org.eclipse.microprofile.concurrent.spi.ThreadContextProvider, is the first point of interaction between the MicroProfile Concurrency implementation and a thread context provider. This interface is the means by which the MicroProfile Concurrency implementation requests the capturing of a particular context type from the current thread. It also provides a way to obtain a snapshot of empty/cleared context of this type and identifies the name by which the user refers to this context type when configuring a ManagedExecutor or ThreadContext.
+
+[source, java]
+----
+package org.eclipse.microprofile.example.context.priority;
+
+import java.util.Map;
+import org.eclipse.microprofile.concurrent.spi.ThreadContextProvider;
+import org.eclipse.microprofile.concurrent.spi.ThreadContextSnapshot;
+
+public class ThreadPriorityContextProvider implements ThreadContextProvider {
+    public String getThreadContextType() {
+        return "ThreadPriority";
+    }
+
+    public ThreadContextSnapshot currentContext(Map<String, String> props) {
+        return new ThreadPrioritySnapshot(Thread.currentThread().getPriority());
+    }
+
+    public ThreadContextSnapshot clearedContext(Map<String, String> props) {
+        return new ThreadPrioritySnapshot(Thread.NORM_PRIORITY);
+    }
+}
+----
+
+==== ThreadContextSnapshot
+
+The interface, org.eclipse.microprofile.concurrent.spi.ThreadContextSnapshot, represents an immutable snapshot of thread context. The MicroProfile Concurrency implementation can request the context represented by this snapshot to be applied to any number of threads by invoking the the begin method. The begin method must save the previous context of the thread so that it can return an instance of org.eclipse.microprofile.concurrent.spi.ThreadContextController for one-time use by the MicroProfile Concurrency implementation to restore the previous context after the context represented by the snapshot is no longer needed on the thread.
```

Comment:
`represents an immutable snapshot of thread context`

Not sure I understand this correctly, does it mean you assume you can use the same context repeatedly in several threads and every time you do that, all the threads will have the same initial state? E.g. that the actual context doesn't change?
This won't really be true for CDI and I am pretty sure transactions will be the same. Take transactions for starters, if you have a running tx you want to propagate, you capture the context and offload the execution elsewhere. However, once done this is done, tx is no longer running. How do you aim to propagate the same context repeatedly in another thread ("some time later")?
In CDI, we can propagate beans as in take what is in current context (talking CDI contexts now) and feed those instances into new thread. But those are ultimately the same beans so you can change their state which will be reflected in any further usage of that snapshot.

Reply:
No, the intent when describing it as immutable was just to discourage the storing of other state (specifically, information about the previous context to restore) within a ThreadContextSnapshot instance because that would prevent the proper application/restore of context when reused for multiple threads.  Another way to think of it is that ThreadContextSnapshot is a pointer to thread context, for example, a transaction id. Even though the transaction might itself be in progress at one time, and at a later time has been committed or rolled back, the transaction id remains the same.
I'll work on coming up with a different wording for this, because we don't want to cause any confusion to context providers.